### PR TITLE
sentinel-terminated pointers

### PR DESCRIPTION
**Pull request by andrewrk**
This implements #265.

 * [x] implement null terminated pointers
 * [x] make string literals null terminated pointers to arrays instead of array values
 * [x] delete C string literals and multiline C string literals from the language
 * [x] go ahead and do #3731
 * [x] get behavior tests passing
 * [x] get std lib tests passing
 * [x] self-hosted / zig fmt support for the new syntax
 * [ ] ~slicing sentinel terminated array or pointer with `..` as end index should give a sentinel-terminated slice~  #3766
 * [x] replace syntax for `[N]null T`, `[]null T`, and `[*]null T`, with the syntax from #3731
 * add implicit casting tests
   - [x] `[:x]T` to `[]T`
   - [x] `[*:x]T` to `[*]T`
   - [x] `[N:x]T` to `[N]T`
   - [x] `*[N:x]T` to `*[N]T`
   - [x] `[:x]T` to `[*:x]T`
   - [x] ~`[N:x]T` to `[:x]const T`~
   - [x] ~`[N:x]T` to `[*:x]const T`~
 * [x] add additional tests for #3731
 * [x] Take advantage of null terminated pointers in the standard library (grep for `#265`). (follow-up issue: #3767)
 * [x] Get the rest of the tests passing
 * [x] Add documentation to the language reference
 * [ ] ~Look for regressions in zig community members' actively maintained projects~ there are already recent regressions such as #3692 which are unrelated to this branch, so I'll be doing this in master branch in the upcoming week or two. Hold on to your butts.
 * [ ] ~Remove type coercion from array values which do not have static lifetimes to references~ #3768
 * [ ] ~when comptime len index, reading gives comptime value. writing asserts the value is the sentinel, and is no-op~ https://github.com/ziglang/zig/issues/3731#issuecomment-556436531
 * [x] compile error for sentinel mismatch when type coercing

### CODE REVIEW

**Comment by daurnimator** [comment]
>     * make string literals null terminated pointers to arrays instead of array values

Does this mean that you can't do `"hello".len` any more? does that end up invoking a comptime `strlen`?

What does this mean for concatenating two strings?
```zig
const a = "foo";
const b = "bar";
const c = a ++ b; // is this null terminated? do I end up with a null in the middle?
```

> `MakeMemNoAccess = valgrind.ToolBase("MC".*),`

Are we going to end up with `.*` everywhere? I'm not entirely clear what the `.*` implies here (it's a pointer dereference

----- 
idea: instead of `[*]null u8`, `[*]sentinel(0) u8`. It's not obvious that `null` means `0`: no where else in zig do we call 0 "null". We could open things up to allow arbitrary values in the future

**Comment by Rocknest** [comment]
This implementation does not seem quite right. It gives impression of a hack. Not sure. 

Maybe there should be a type `comptime_string` that can coerce to `[*]u8`, `[*]null u8` and slices. 

Also i like @daurnimator's idea: `[*]sentinel(0) u8` or `[*]zero_sentinel u8`.

**Comment by andrewrk** [comment]
> Does this mean that you can't do `"hello".len` any more? does that end up invoking a comptime `strlen`?

This code works in master branch and in this branch: `"hello".len == 5`. No there is no comptime strlen. String literals have both a comptime known length and are also null terminated. They safely cast to both `[]const u8` and `[*]null const u8`.

> What does this mean for concatenating two strings?

In the expression `a ++ b`, if either of `a` or `b` is a null terminated array, then the result is a null terminated array. It works with single-item pointers to arrays as well. If either is a single-item pointer to an array then the result is a single-item pointer to an array.

One of the main benefits of having null termination in the type system is that operations such as `++`, `**`, and `len` property can work correctly.

> Are we going to end up with `.*` everywhere? I'm not entirely clear what the `.*` implies here (it's a pointer dereference

`typeOf("MC") == *const [2]null u8`. So, dereferencing it gives you an array value of type `[2]null u8`. The question of whether to allow a type coercion from `*[N]T` to `[N]T` is a separate issue.

> idea: instead of `[*]null u8`, `[*]sentinel(0) u8`. It's not obvious that `null` means `0`: no where else in zig do we call 0 "null". We could open things up to allow arbitrary values in the future

Maybe you are onto something here, but please open a separate proposal. This is an implementation of the accepted-for-over-two-years proposal #265.

**Comment by daurnimator** [comment]
> Maybe you are onto something here, but please open a separate proposal. This is an implementation of the accepted-for-over-two-years proposal #265.

Created as #3731

However now I'm curious: for non-pointer non-integer elements, what does null mean? e.g. `[*]null enum{Foo, Bar}`

**Comment by andrewrk** [comment]
> However now I'm curious: for non-pointer non-integer elements, what does null mean? e.g. `[*]null enum{Foo, Bar}`

Not allowed, with the current implementation. error: type "E" has no null value, cannot be used in null terminated pointer.

**Comment by daurnimator** [comment]
A good update to make would be for `std.mem.toSlice` and `std.mem.toSliceConst` to return null terminated slices.

**Comment by daurnimator** [comment]
> A good update to make would be for `std.mem.toSlice` and `std.mem.toSliceConst` to return null terminated slices.

Attempting this I got:
```
broken LLVM module found: Call parameter type does not match function signature!
  %18 = getelementptr inbounds { %"[]u8", i16 }, { %"[]u8", i16 }* %0, i32 0, i32 0, !dbg !126586
 %"[:0]u8"*  call fastcc void @mem.toSlice(%"[]u8"* sret %18, i8* %16), !dbg !126586
Call parameter type does not match function signature!
  %name = alloca %"[:0]u8", align 8
 %"[]u8"*  %48 = call fastcc i1 @mem.eql(%"[:0]u8"* %name, %"[]u8"* @5875), !dbg !211774
Call parameter type does not match function signature!
  %name = alloca %"[:0]u8", align 8
 %"[]u8"*  %84 = call fastcc i1 @mem.eql(%"[:0]u8"* %name, %"[]u8"* @5874), !dbg !211810
Call parameter type does not match function signature!
  %hostname = alloca %"[:0]u8", align 8
 %"[]u8"*  call fastcc void @mem.copy(%"[]u8"* %9, %"[:0]u8"* %hostname), !dbg !251182
Call parameter type does not match function signature!
  %27 = getelementptr inbounds %"?[]const u8", %"?[]const u8"* %0, i32 0, i32 0, !dbg !284345
 %"[:0]u8"*  call fastcc void @mem.toSlice(%"[]u8"* sret %27, i8* %25), !dbg !284345
```

File: test/compile_errors.zig [hunk 1]
```diff
@@ -2,6 +2,21 @@ const tests = @import("tests.zig");
...
+        "incompatible pointer sentinels",
+        \\export fn entry1(ptr: [*:255]u8) [*:0]u8 {
+        \\    return ptr;
+        \\}
+        \\export fn entry2(ptr: [*]u8) [*:0]u8 {
+        \\    return ptr;
+        \\}
+    ,
+        "tmp.zig:2:5: error: expected type '[*:0]u8', found '[*:255]u8'",
+        "tmp.zig:2:5: note: destination pointer requires a terminating '0' sentinel value, but source pointer has a terminating '255' sentinel value",
```
Comment (daurnimator) [inline test/compile_errors.zig]: s/sentinel value/sentinel/?

**Pull request merged**